[quant] Input-Weight Equalization - selective equalization

angelayi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #61916

Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
layer_to_sqnr_dict = get_layer_sqnr_dict(float_model, equalized_model, input)
eq_qconfig_dict = get_equalization_qconfig_dict(layer_to_sqnr_dict, equalized_model, num_layers_to_equalize)
```
Script for an example of running these functions is available in the diff [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950).

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

angelayi committed:
[quant] Input-Weight Equalization - selective equalization

Summary: Functions used to run selective equalization based on the SQNR
obtained from running the Numeric Suite. After running the Numeric Suite
between the equalized and float model, we will get the SQNR between the
two models and construct an equalization_qconfig_dict that specifies to
only equalize the layers with the highest quantization errors.

How to run:
```
layer_to_sqnr_dict = get_layer_sqnr_dict(float_model, equalized_model, input)
eq_qconfig_dict = get_equalization_qconfig_dict(layer_to_sqnr_dict, equalized_model, num_layers_to_equalize)

prepared = prepare_fx(float_model, qconfig_dict, eq_qconfig_dict)
...
```

Test Plan: python test/test_quantization.py
TestEqualizeFx.test_selective_equalization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors.

How to run:
```
layer_to_sqnr_dict = get_layer_sqnr_dict(float_model, equalized_model, input)
eq_qconfig_dict = get_equalization_qconfig_dict(layer_to_sqnr_dict, equalized_model, num_layers_to_equalize)

prepared = prepare_fx(float_model, qconfig_dict, eq_qconfig_dict)
...
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29796950).

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -698,3 +703,100 @@ def _convert_equalization_ref(model: GraphModule):
+# Functions for running the equalized model on the Numeric Suite              #
+###############################################################################
+
+def get_layer_sqnr_dict(model_a: nn.Module, model_b: nn.Module, x: Any) -> Dict[str, float]:
```
Can `_get_layer_sqnr_dict` and `_get_equalization_qconfig_dict` be helper functions instead and we have a single top level API for selective_equalization with numeric suite which calls them?

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors.

How to run:
```
layer_to_sqnr_dict = get_layer_sqnr_dict(float_model, equalized_model, input)
eq_qconfig_dict = get_equalization_qconfig_dict(layer_to_sqnr_dict, equalized_model, num_layers_to_equalize)

prepared = prepare_fx(float_model, qconfig_dict, eq_qconfig_dict)
...
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
selectively_equalized_model = selctive_equalization_with_numeric_suite(float_model, qconfig_dict, input, num_layers_to_equalize)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

supriyar commented on torch/quantization/fx/_equalize.py:
```diff
@@ -698,3 +705,155 @@ def _convert_equalization_ref(model: GraphModule):
+    prepare_custom_config_dict: Dict[str, Any] = None,
+    convert_custom_config_dict: Dict[str, Any] = None,
+) -> QuantizedGraphModule:
+    """ Selectively equalizes the given float model with the following steps:
```
I think we should update the comment and code to say this is equalized + quantized output

supriyar reviewed (approved):
Looks good, test failure seems related. You probably need to add the fbgemm decorator

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
selectively_equalized_model = selctive_equalization_with_numeric_suite(float_model, qconfig_dict, input, num_layers_to_equalize)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29796950).

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
selectively_equalized_model = selctive_equalization_with_numeric_suite(float_model, qconfig_dict, input, num_layers_to_equalize)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29796950).

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
selectively_equalized_model = selctive_equalization_with_numeric_suite(float_model, qconfig_dict, input, num_layers_to_equalize)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29796950).

raghuramank100 reviewed (changes requested):
I think while this is a good example to test, this change should not be part of the repo as it is more of a top level example which involves calling the numeric suite APIs.  It feels like instead we should be 
1. Making modifications to numeric suite code to add a small utility function to return a SQNR dict keyed by layer names (Fully qualified names).
2. Fix the numeric suite so that it can handle the equalization scaling for the activations.
3. Directly map from the SQNR Dict to the equalization config dict.
The rest of the code can be a notebook showing how to call these lower level APIs, so that we avoid having complex APIs.

raghuramank100 commented on torch/quantization/fx/_equalize.py:
```diff
@@ -698,3 +705,157 @@ def _convert_equalization_ref(model: GraphModule):
+    layers_to_equalize = layer_sqnr_sorted[:num_layers_to_equalize]
+
+    # Trace the model so that we can find the accurate node_name_to_scope
+    from ..quantize_fx import QuantizationTracer
```
Is it possible to avoid the tracer here and directly go from layers_to_equalize to the equalization_qconfig_dict?. Avoiding the need to have the model as an argument would make this API cleaner.

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
selectively_equalized_model = selctive_equalization_with_numeric_suite(float_model, qconfig_dict, input, num_layers_to_equalize)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29796950).

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
selectively_equalized_model = selctive_equalization_with_numeric_suite(float_model, qconfig_dict, input, num_layers_to_equalize)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29796950).

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
selectively_equalized_model = selctive_equalization_with_numeric_suite(float_model, qconfig_dict, input, num_layers_to_equalize)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29796950).

angelayi committed:
Update on "[quant] Input-Weight Equalization - selective equalization"


Summary: Functions used to run selective equalization based on the SQNR obtained from running the Numeric Suite. After running the Numeric Suite between the equalized and float model, we will get the SQNR between the two models and construct an equalization_qconfig_dict that specifies to only equalize the layers with the highest quantization errors. Then, we will run equalization on only those layers.

How to run:
```
selectively_equalized_model = selctive_equalization_with_numeric_suite(float_model, qconfig_dict, input, num_layers_to_equalize)
```

Test Plan: `python test/test_quantization.py TestEqualizeFx.test_selective_equalization`

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29796950](https://our.internmc.facebook.com/intern/diff/D29796950)

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29796950).
